saq/queue/postgres.py

```diff
@@ -0,0 +1,334 @@
+"""
+Postgres Queue
+"""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+import time
+import typing as t
+
+from psycopg.types.json import Jsonb
+from psycopg_pool import AsyncNullConnectionPool, AsyncConnectionPool
+
+from saq.job import (
+    Job,
+    Status,
+)
+from saq.queue.base import Queue, logger
+from saq.utils import now
+
+if t.TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from saq.types import (
+        CountKind,
+        ListenCallback,
+        DumpType,
+        LoadType,
+        QueueInfo,
+        QueueStats,
+    )
+
+JOBS_TABLE = "saq_jobs"
+
+CREATE_JOBS_TABLE = """
+CREATE TABLE IF NOT EXISTS saq_jobs (
+    key TEXT PRIMARY KEY,
+    queued BIGINT NOT NULL DEFAULT EXTRACT(EPOCH FROM now()),
+    function TEXT,
+    job JSONB,
+    queue TEXT,
+    status TEXT,
+    scheduled BIGINT
+);
+"""
+
+
+class PostgresQueue(Queue):
+    """
+    Queue is used to interact with redis.
+
+    Args:
+        pool: instance of psycopg_pool.AsyncConnectionPool
+        name: name of the queue (default "default")
+        dump: lambda that takes a dictionary and outputs bytes (default `json.dumps`)
+        load: lambda that takes str or bytes and outputs a python dictionary (default `json.loads`)
+        max_size: maximum pool size. (default 5)
+            If greater than 0, this limits the maximum number of connections of the connection pool.
+            Otherwise, the pool will have no limit.
+    """
+
+    @classmethod
+    def from_url(cls: type[PostgresQueue], url: str, **kwargs: t.Any) -> PostgresQueue:
+        """Create a queue from a postgres url."""
+        return cls(AsyncNullConnectionPool(url, open=False), **kwargs)
+
+    def __init__(
+        self,
+        pool: AsyncConnectionPool,
+        name: str = "default",
+        dump: DumpType | None = None,
+        load: LoadType | None = None,
+        max_size: int = 5,
+    ) -> None:
+        super().__init__(name=name, dump=dump, load=load)
+
+        self.pool = pool
+        self.max_size = max_size
+
+    async def connect(self) -> None:
+        await self.pool.open()
+        await self.pool.resize(min_size=0, max_size=self.max_size)
```

Comment:
what does resize min_size=0 do?

Reply:
For AsyncNullConnectionPool, min_size is always 0 since it doesn't maintain any connections. I've switched the queue to use an AsyncConnectionPool.